framework/db/src/com/cloud/utils/db/SqlGenerator.java

```diff
@@ -680,4 +680,21 @@ public String buildDistinctIdSql() {
 
         return sql.append("SELECT DISTINCT id FROM ").append(buildTableReferences()).append(" WHERE ").append(buildDiscriminatorClause().first()).toString();
     }
+
+    public String buildDistinctSql(String[] distinctColumnNames) {
+        StringBuilder sbColumn = new StringBuilder();
+
+        if(distinctColumnNames != null && distinctColumnNames.length > 0) {
+            for (String columnName : distinctColumnNames) {
+                sbColumn.append(columnName).append(", ");
+            }
+            sbColumn.delete(sbColumn.length() - 2, sbColumn.length());
+        }
+        else{
```

Comment:
Can you move the else to look like `} else {`

Reply:
Fixed